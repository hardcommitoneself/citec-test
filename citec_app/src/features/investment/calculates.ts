// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
export const calculateProgressionSavingFeesAndTaxes = (
   years: number,
   type: string,
   investment: number,
   setInvestmentAcumulative: any,
   setSerieInvestmentPerYearWithSavings: any,
   setSerieInvestmentPerYearWithoutSavings: any
) => {
   let acumulativeSavings = 0
   let acumulativeAmount = investment
   let acumulativeAmount10Percent = investment
   const acumulativePerYearWithSavings = []
   const acumulativePerYearWithoutSavings = []
   acumulativePerYearWithSavings[0] = investment
   acumulativePerYearWithoutSavings[0] = investment

   for (let i = 1; i <= years; i++) {

      acumulativePerYearWithSavings[i] = acumulativePerYearWithSavings[i - 1] * 1.11
      acumulativePerYearWithoutSavings[i] = acumulativePerYearWithoutSavings[i - 1] * 1.08
      acumulativeAmount *= 1.11
      acumulativeAmount10Percent *= 1.1

   }

   if (type == 'fees') {
      acumulativeSavings = acumulativeAmount10Percent - acumulativePerYearWithoutSavings[years]
   } else {
      acumulativeSavings = acumulativeAmount - acumulativeAmount10Percent
   }
   setInvestmentAcumulative(acumulativeAmount)
   setSerieInvestmentPerYearWithSavings(acumulativePerYearWithSavings)
   setSerieInvestmentPerYearWithoutSavings(acumulativePerYearWithoutSavings)
   return acumulativeSavings
}