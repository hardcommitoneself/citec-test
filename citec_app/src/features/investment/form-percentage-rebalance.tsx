// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import FormControl from '@mui/material/FormControl'
import FormControlLabel from '@mui/material/FormControlLabel'
import Radio from '@mui/material/Radio'
import RadioGroup from '@mui/material/RadioGroup'

import { fetchDataFromCitecApi } from 'features/utils/api/citec-api'
import { setPositionSelected } from 'store/slices/tax-harvesting'
import { setHold } from 'store/slices/rebalance'
import { useAppSelector } from 'store/hooks'
import {
   check100Percentage,
   check25or50Percentage,
   checkNegativePositions,
} from 'features/utils/percentage-rebalance'
import { useState } from 'react'

interface TaxHarvestingResp {
   ticker: string[]
   holds: number[]
}

export const FormPercentageRebalance = () => {
   const [value, setValue] = useState(0)
   const portfolio = useAppSelector((state) => state.portfolio)
   const portfolioOptimized = useAppSelector((state) => state.portfolio_optimized)
   const holds = useAppSelector((state) => state.rebalance.hold)

   const dispatch = useDispatch()

   const handleSelectPercentaje = async ({
      target,
   }: React.ChangeEvent<HTMLInputElement>) => {
      const selectedValue = Number(target.value)
      const { outstanding_balance, unrealized_capital_gains, ticker } = portfolio

      setValue(selectedValue)

      const { isSomeNegativePosition, negativeUnrealized } = checkNegativePositions({
         outstanding_balance,
         ticker,
         unrealized_capital_gains,
      })

      if (!isSomeNegativePosition) return

      if (selectedValue === 0) return

      if (selectedValue === 100) {
         const updatedHolds = check100Percentage({
            holds,
            negativeUnrealized,
            ticker,
         })

         dispatch(
            setPositionSelected({
               arrayPositionSelected: negativeUnrealized,
            })
         )

         dispatch(setHold(updatedHolds))

         return
      }

      const {
         benchmark,
         portfolio_id,
         creation_date,
         portfolio: portfolioName,
         optimization_status,
      } = portfolioOptimized

      const payload = {
         portfolio_id,
         portfolio: portfolioName,
         creation_date,
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
         optimization_status,
      }

      const { data } = await fetchDataFromCitecApi<TaxHarvestingResp>({
         endpoint: '/ai/portfolio/tax_harvesting/',
         method: 'POST',
         params: {
            pct_losses: `.${selectedValue}`,
         },
         payload,
      })

      const indexCash = ticker.indexOf('$CASH')

      data.holds[indexCash] = 0

      dispatch(setHold(data.holds))

      const negativeUnrealizedPercentage = check25or50Percentage({
         holds: data.holds,
         outstanding_balance,
         ticker,
         unrealized_capital_gains,
      })

      dispatch(
         setPositionSelected({
            arrayPositionSelected: negativeUnrealizedPercentage,
         })
      )
   }

   const formValues = [0, 25, 50, 100]

   return (
      <FormControl>
         <RadioGroup
            aria-labelledby="demo-radio-buttons-group-label"
            name="radio-buttons-group"
            onChange={handleSelectPercentaje}
            value={value}
         >
            {formValues.map((val) => (
               <FormControlLabel
                  key={val}
                  value={val}
                  control={<Radio />}
                  label={
                     val === 0 ? 'Harvest losses manualy' : `${val}% of tax losses`
                  }
               />
            ))}
         </RadioGroup>
      </FormControl>
   )
}
