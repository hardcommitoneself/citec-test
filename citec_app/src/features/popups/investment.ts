// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import Swal from 'sweetalert2'
import { commonsConfig } from './commons-config'

type NamesPopup =
   | 'investment'
   | 'potentialSavings'
   | 'chartSimulations'

const investment = `<div class="align-left">
How much money you should Investment in <u>US Equities?</u>
<br>
&bull; Define how much money you want to allocate to <u>liquid assets</u>
<br>
&bull; Define your <u>Asset Allocation</u> based on your needs and risk tolerance (Typicaly US Equities goes from 40% to 70% of the portfolio)
   </div>`

const potentialSavings = `<div class="align-left">
You can save thousands of dollars by directly managing your money and investing directly in stocks and bonds:
<br>
&bull; <u>Advisory fees</u>: Eliminate fees payed to Wealth Managers (~1.5% a year)
<br>
&bull; <u>Structoring fees</u>: Eliminate fees payed to Asset Managers (~0.5% a year)
<br>
&bull; <u>Tax Alpha</u>: reduce overall tax bill by systematically harvest tax lossess (~1% a year)
   </div>`

const chartSimulations = `<div class="align-left">
<b>Key assumptions:</b>
<br>
&bull; <u>US Equity Market Returns </u>(~10.5% a year)
<br>
&bull; <u>Total Fees</u> including Advisory, Structoring, Trading (~2% a year)
<br>
&bull; <u>Tax Alpha</u> (~1.0%)
   </div>`


export const renderPopup = (name: NamesPopup) => {
   let html
   switch (name) {
      case 'investment':
         html = investment
         break
      case 'potentialSavings':
         html = potentialSavings
         break
      case 'chartSimulations':
         html = chartSimulations
         break
      default:
         html = ''
         break
   }

   Swal.fire({
      html,
      ...commonsConfig,
   })
}
