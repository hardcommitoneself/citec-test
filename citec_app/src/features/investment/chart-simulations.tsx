// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { formatToMoney } from 'features/utils/format-to-money'
import ReactApexChart from 'react-apexcharts'
import { renderPopup } from 'features/popups/investment'
const Group = require('assets/img/group.png')

interface ChartSimulationProps {
   years: number
   setYears: React.Dispatch<React.SetStateAction<number>>
   acumulative: number
   serieInvestmentPerYearWithSavings: Array<number>
   serieInvestmentPerYearWithoutSavings: Array<number>
}

export const ChartSimulation = ({
   years,
   setYears,
   acumulative,
   serieInvestmentPerYearWithSavings,
   serieInvestmentPerYearWithoutSavings,
}: ChartSimulationProps) => {
   let customCategoriesX = [] as Array<string>

   const series = [
      {
         name: 'Performance with savings',
         data: serieInvestmentPerYearWithSavings,
      },
      {
         name: 'Performance without savings',
         data: serieInvestmentPerYearWithoutSavings,
      },
   ]

   if (series.length == 6) {
      customCategoriesX = ['0', '1', '2', '3', '4', '5']
   } else if (series.length == 11) {
      customCategoriesX = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
   } else {
      customCategoriesX = [
         '0',
         '1',
         '2',
         '3',
         '4',
         '5',
         '6',
         '7',
         '8',
         '9',
         '10',
         '11',
         '12',
         '13',
         '14',
         '15',
         '16',
         '17',
         '18',
         '19',
         '20',
      ]
   }

   const options = {
      colors: ['#20BBB7', '#38B2ED'],
      chart: {
         type: 'area' as const,
         height: 350,
         zoom: {
            autoScaleYaxis: true,
         },
      },
      dataLabels: {
         enabled: false,
      },
      stroke: {
         curve: 'straight' as const,
         width: 2.5,
      },
      legend: {
         //https://apexcharts.com/docs/options/legend/
         show: true,
         position: 'bottom' as const,
         floating: true,
         horizontalAlign: 'right' as const,
         offsetY: 8,
         fontSize: '13px',
      },
      xaxis: {
         categories: customCategoriesX,
         title: {
            text: 'Years',
            style: {
               cssClass: 'apexcharts-yaxis-label-investment',
            },
         },
      },
      yaxis: {
         tickAmount: 4, // apexcharts.com/docs/options/yaxis/
         decimalsInFloat: 0,
         overwriteCategories: undefined,
         labels: {
            style: {
               cssClass: 'apexcharts-yaxis-label-investment',
            },
            /**
             * Allows users to apply a custom formatter function to x-axis labels.
             *
             * @param { String } value - The default value generated
             * @param { Number } timestamp - In a datetime series, this is the raw timestamp
             * @param { object } contains dateFormatter for datetime x-axis
             */
            formatter: function (value: any) {
               return formatToMoney(Math.trunc(value))
            },
         },
      },
      tooltip: {
         x: {
            show: false,
         },
      },
   }

   return (
      <div className="box chart_box">
         <div className="top_content">
            <div className="left_side ">
               <div className="row-chart-investment">
                  <h1>Simulations</h1>
               </div>
            </div>
            <div className="right_side">
               <div className="row-chart-investment">
                  <h1 className="investement-expected-value">
                     Expected value:{' '}
                     <span> {formatToMoney(Math.trunc(acumulative))}</span>
                  </h1>
               </div>
            </div>
         </div>

         <div className="top_content">
            <div className="left_side" style={{ paddingTop: '20px' }}>
               <div className="investment-text-after-returns">
                  <span className="flex items-center text-[#0BACA9]">
                     <p className="mr-2">After tax returns simulations</p>
                     <img
                        className="moreinfo"
                        src={Group}
                        alt=""
                        onClick={() => {
                           renderPopup('chartSimulations')
                        }}
                        style={{ marginLeft: '2px' }}
                     />
                  </span>
               </div>
            </div>
            <div className="right_side">
               <div className="buttons_small_area items-end">
                  <button
                     className={years === 5 ? 'active' : ''}
                     onClick={() => setYears(5)}
                     data-testid="5Y"
                  >
                     5Y
                  </button>
                  <button
                     className={years === 10 ? 'active' : ''}
                     onClick={() => setYears(10)}
                     data-testid="10Y"
                  >
                     10Y
                  </button>
                  <button
                     data-testid="20Y"
                     className={years === 20 ? 'active' : ''}
                     onClick={() => setYears(20)}
                  >
                     20Y
                  </button>
               </div>
            </div>
         </div>

         <div className="pt-2">
            <ReactApexChart
               options={options}
               series={series}
               type="area"
               height={350}
            />
         </div>
      </div>
   )
}
