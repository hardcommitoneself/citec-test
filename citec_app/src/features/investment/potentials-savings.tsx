// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { formatToMoney } from 'features/utils/format-to-money'
import { renderPopup } from 'features/popups/investment'

const Group = require('assets/img/group.png')

const savings = require('assets/img/savings.png')

interface PotentialsSavingsProps {
   fees: number
   taxes: number
   years: number
   serieInvestmentPerYearWithSavings: Array<number>
   serieInvestmentPerYearWithoutSavings: Array<number>
}

export const PotentialsSavings = ({
   fees,
   taxes,
   years,
   serieInvestmentPerYearWithSavings,
   serieInvestmentPerYearWithoutSavings
}: PotentialsSavingsProps) => {
   const sizeArray = serieInvestmentPerYearWithSavings.length
   const totalSavings = (
      serieInvestmentPerYearWithSavings[sizeArray - 1] -
      serieInvestmentPerYearWithoutSavings[sizeArray - 1]
   )

   return (
      <div className="box">
         <img src={savings} alt="" />
         <div className="bottom_area">
            <div className="row_custom" style={{ width: '100%' }}>
               <h3 className="potential_savings">
                  Potential savings in {years} years
                  <img
                     className="moreinfo"
                     src={Group}
                     alt=""
                     onClick={() => {
                        renderPopup('potentialSavings')
                     }}
                     style={{ marginLeft: '3px' }}
                  />
               </h3>
            </div>
            <div className="row_custom">
               <h1>Savings in fees:</h1>
               <p>{formatToMoney(Math.trunc(fees))}</p>
            </div>
            <div className="row_custom" style={{ paddingTop: '0px' }}>
               <h1>Savings in taxes:</h1>
               <p>{formatToMoney(Math.trunc(taxes))}</p>
            </div>
            <div className="footer_row row_custom">
               <h1>Total savings:</h1>
               <p>{formatToMoney(Math.trunc(totalSavings))}</p>
            </div>
         </div>
      </div>
   )
}
