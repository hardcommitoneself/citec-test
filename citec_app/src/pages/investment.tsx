// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import { Button, Slider } from '@mui/material'
import { useEffect, useState } from 'react'
import NumberFormat from 'react-number-format'
import { PageContainer } from 'features/ui/page-container'
import { setCash, setUpdatedCash } from 'store/slices/portfolio'
import { renderPopup } from 'features/popups/investment'
import { useAppSelector } from 'store/hooks'
import {
   calculateProgressionSavingFeesAndTaxes,
   ChartSimulation,
   FormPercentageRebalance,
   PotentialsSavings,
} from 'features/investment'
import {
   useAppNavigate,
   useHandlerTaxHarvesting,
   useValidateNumberStocks,
   useValidateSoldPosition,
   useValidationPorfolioLoaded,
} from 'features/utils/hooks'

const Group = require('assets/img/group.png')

export const Investment = () => {
   const dispatch = useDispatch()
   const navigate = useAppNavigate()
   const { soldCashOnTaxharvesting } = useValidateSoldPosition()
   const { checkValidStocks } = useValidateNumberStocks()
   useHandlerTaxHarvesting()

   const [investmentAcumulative, setInvestmentAcumulative] = useState(1000000)
   const [serieInvestmentPerYearWithSavings, setSerieInvestmentPerYearWithSavings] =
      useState([])
   const [
      serieInvestmentPerYearWithoutSavings,
      setSerieInvestmentPerYearWithoutSavings,
   ] = useState([])
   const [savings, setSavings] = useState({
      fees: 200000,
      taxes: 400000,
   })
   const [investment, setInvestment] = useState(1000000)
   const [years, setYears] = useState(10)
   const workflow = useAppSelector((state) => state.workflow.type)
   const MAX_VAL = 10000000

   const { isPortfolioLoaded } = useValidationPorfolioLoaded()

   const portfolioTickers = useAppSelector((state) => state.portfolio.ticker)

   const indexCash = portfolioTickers.indexOf('$CASH')
   const isCashOnPortfolio = indexCash >= 0

   const setPortfolioCash = () => {
      const cash = {
         outstanding_balance: investment,
         unrealized_capital_gains: 0,
      }

      dispatch(setUpdatedCash(cash))

      soldCashOnTaxharvesting(investment)
   }

   const investmentCash = useAppSelector(
      (state) => state.portfolio.outstanding_balance[indexCash]
   )

   const handleInvestmentChange = (value: number) => {
      setSavings({
         fees: calculateProgressionSavingFeesAndTaxes(
            years,
            'fees',
            investment,
            setInvestmentAcumulative,
            setSerieInvestmentPerYearWithSavings,
            setSerieInvestmentPerYearWithoutSavings
         ),
         taxes: calculateProgressionSavingFeesAndTaxes(
            years,
            'taxes',
            investment,
            setInvestmentAcumulative,
            setSerieInvestmentPerYearWithSavings,
            setSerieInvestmentPerYearWithoutSavings
         ),
      })
      setInvestment(value)
   }

   useEffect(() => {
      if (workflow === 'new') {
         handleInvestmentChange(investment)
      }
   }, [years])

   useEffect(() => {
      if (isCashOnPortfolio) {
         setInvestment(investmentCash)
      } else {
         dispatch(
            setCash({
               outstanding_balance: investment,
               unrealized_capital_gains: 0,
            })
         )
      }
      if (!isPortfolioLoaded) setPortfolioCash()

      soldCashOnTaxharvesting(isCashOnPortfolio ? investmentCash : investment)
   }, [])

   useEffect(() => {
      checkValidStocks()
   }, [investmentCash])

   return (
      <PageContainer>
         <div className="dashboard_wrapper">
            <div className="InvestmentBody dashboard_body">
               <div className="top_boxes_area_investment">
                  <div className={workflow == 'new' ? 'box' : 'box100'}>
                     {workflow == 'new' ? (
                        <h2>How much money you want to invest? </h2>
                     ) : (
                        <h2>Do you want to add more cash? </h2>
                     )}
                     <div className="inner_box">
                        <div className="top_area text-[#0BACA9]">
                           <h1>Investment in cash</h1>
                           <img
                              className="moreinfo"
                              src={Group}
                              alt=""
                              onClick={() => {
                                 renderPopup('investment')
                              }}
                              style={{ marginLeft: '3px' }}
                           />
                        </div>
                        <NumberFormat
                           value={investment}
                           thousandSeparator={true}
                           min={0}
                           maxLength={11}
                           isAllowed={({ value }) => {
                              if (Number(value) <= MAX_VAL) return true
                              return false
                           }}
                           onValueChange={({ value }) => {
                              handleInvestmentChange(Number(value))
                           }}
                           onBlur={setPortfolioCash}
                           prefix="$"
                           className="price_area w-full"
                           id="investment-input"
                        />

                        <Slider
                           valueLabelDisplay="auto"
                           onChange={(_, newValue) => {
                              setInvestment(newValue as number)
                           }}
                           onChangeCommitted={setPortfolioCash}
                           value={investment}
                           min={0}
                           max={10000000}
                        />
                     </div>
                  </div>
                  {workflow === 'new' && (
                     <PotentialsSavings
                        {...savings}
                        years={years}
                        serieInvestmentPerYearWithSavings={
                           serieInvestmentPerYearWithSavings
                        }
                        serieInvestmentPerYearWithoutSavings={
                           serieInvestmentPerYearWithoutSavings
                        }
                     />
                  )}
               </div>

               {workflow === 'new' ? (
                  <ChartSimulation
                     years={years}
                     setYears={setYears}
                     acumulative={investmentAcumulative}
                     serieInvestmentPerYearWithSavings={
                        serieInvestmentPerYearWithSavings
                     }
                     serieInvestmentPerYearWithoutSavings={
                        serieInvestmentPerYearWithoutSavings
                     }
                  />
               ) : (
                  <div className="box chart_box">
                     <h2 className="mb-[10px]">
                        How do you want to rebalance the portfolio?
                     </h2>
                     <FormPercentageRebalance />
                  </div>
               )}

               <div className="buttons_wrapper">
                  <Button
                     color="secondary"
                     onClick={() =>
                        navigate(
                           workflow === 'new' ? '/workflow' : '/portfolio-analysis'
                        )
                     }
                  >
                     Back
                  </Button>
                  <Button
                     onClick={() =>
                        navigate(
                           workflow === 'new' ? '/criteria' : '/tax-harvesting'
                        )
                     }
                  >
                     Next
                  </Button>
               </div>
            </div>
         </div>
      </PageContainer>
   )
}
