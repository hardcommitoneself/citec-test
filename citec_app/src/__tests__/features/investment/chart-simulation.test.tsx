// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { fireEvent, render } from '@testing-library/react'
import { shallow } from 'enzyme'
import { ChartSimulation } from 'features/investment'

jest.mock('react-apexcharts', () => {
   return {
      __esModule: true,
      default: () => {
         return <div />
      },
   }
})

const setYears = jest.fn()
const investmentAcumulative = jest.fn() as any
const serieInvestmentPerYearWithSavings = jest.fn() as any
const serieInvestmentPerYearWithoutSavings = jest.fn() as any

describe('Test at ChartSimulation component', () => {
   test('snapshot and click change year', () => {
      const { container, getByTestId } = render(
         <ChartSimulation
            years={10}
            setYears={setYears}
            acumulative={investmentAcumulative}
            serieInvestmentPerYearWithSavings={serieInvestmentPerYearWithSavings}
            serieInvestmentPerYearWithoutSavings={
               serieInvestmentPerYearWithoutSavings
            }
         />
      )

      fireEvent.click(getByTestId('10Y'))

      expect(container).toMatchSnapshot()
      expect(setYears).toBeCalledTimes(1)
   })

   test('click change year to 5', () => {
      const { getByTestId } = render(
         <ChartSimulation
            years={5}
            setYears={setYears}
            acumulative={investmentAcumulative}
            serieInvestmentPerYearWithSavings={serieInvestmentPerYearWithSavings}
            serieInvestmentPerYearWithoutSavings={
               serieInvestmentPerYearWithoutSavings
            }
         />
      )

      fireEvent.click(getByTestId('5Y'))

      expect(setYears).toBeCalledTimes(1)
   })

   test('click change year to 20', () => {
      const { getByTestId } = render(
         <ChartSimulation
            years={10}
            setYears={setYears}
            acumulative={investmentAcumulative}
            serieInvestmentPerYearWithSavings={serieInvestmentPerYearWithSavings}
            serieInvestmentPerYearWithoutSavings={
               serieInvestmentPerYearWithoutSavings
            }
         />
      )

      fireEvent.click(getByTestId('20Y'))

      expect(setYears).toBeCalledTimes(1)
   })

   test('active class in button 20Y', async () => {
      const wrapper = shallow(
         <ChartSimulation
            years={20}
            setYears={setYears}
            acumulative={investmentAcumulative}
            serieInvestmentPerYearWithSavings={serieInvestmentPerYearWithSavings}
            serieInvestmentPerYearWithoutSavings={
               serieInvestmentPerYearWithoutSavings
            }
         />
      )
      expect(wrapper.find({ 'data-testid': '20Y' }).hasClass('active')).toBeTruthy()
   })
})
