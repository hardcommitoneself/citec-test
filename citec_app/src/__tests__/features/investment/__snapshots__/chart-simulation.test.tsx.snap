// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test at ChartSimulation component snapshot and click change year 1`] = `
<div>
  <div
    class="box chart_box"
  >
    <div
      class="top_content"
    >
      <div
        class="left_side "
      >
        <div
          class="row-chart-investment"
        >
          <h1>
            Simulations
          </h1>
        </div>
      </div>
      <div
        class="right_side"
      >
        <div
          class="row-chart-investment"
        >
          <h1
            class="investement-expected-value"
          >
            Expected value:
             
            <span>
               
              -$ NaN
            </span>
          </h1>
        </div>
      </div>
    </div>
    <div
      class="top_content"
    >
      <div
        class="left_side"
        style="padding-top: 20px;"
      >
        <div
          class="investment-text-after-returns"
        >
          <span
            class="flex items-center text-[#0BACA9]"
          >
            <p
              class="mr-2"
            >
              After tax returns simulations
            </p>
            <img
              alt=""
              class="moreinfo"
              src="group.png"
              style="margin-left: 2px;"
            />
          </span>
        </div>
      </div>
      <div
        class="right_side"
      >
        <div
          class="buttons_small_area items-end"
        >
          <button
            class=""
            data-testid="5Y"
          >
            5Y
          </button>
          <button
            class="active"
            data-testid="10Y"
          >
            10Y
          </button>
          <button
            class=""
            data-testid="20Y"
          >
            20Y
          </button>
        </div>
      </div>
    </div>
    <div
      class="pt-2"
    >
      <div />
    </div>
  </div>
</div>
`;
